packages/replay_bloc/lib/src/replay_bloc.dart

```diff
@@ -102,31 +102,33 @@ mixin ReplayBlocMixin<Event extends ReplayEvent, State> on Bloc<Event, State> {
 
   @override
   void emit(State state) {
-    _changeStack.add(_Change<State>(
-      this.state,
-      () {
-        final event = _Redo();
-        onEvent(event);
-        onTransition(Transition(
-          currentState: this.state,
-          event: event,
-          nextState: state,
-        ));
-        // ignore: invalid_use_of_visible_for_testing_member
-        super.emit(state);
-      },
-      (val) {
-        final event = _Undo();
-        onEvent(event);
-        onTransition(Transition(
-          currentState: this.state,
-          event: event,
-          nextState: val,
-        ));
-        // ignore: invalid_use_of_visible_for_testing_member
-        super.emit(val);
-      },
-    ));
+    if (shouldReplay(this.state)) {
```

Comment:
What if `shouldReplay` evaluates to false when the state changes occurs but at a later time evaluates to true for the same state? As a developer, I would expect that the `ReplayBloc` maintain all states regardless and "skip" certain states that don't meet my criteria rather than permanently excluding states from the change stack. I can try to put together an example of what I'm describing in a bit.

Reply:
Alright, I refactored this to keep all states in the stack and perform the check when you try to `undo`. This way:
1. If there are no states where `shouldReplay` returns true, `canUndo` is false
2. If the latest state doesn't replay, it's popped and we keep trying until one does

Reply:
That sounds good, now I think we're not handling `redo` properly (see my most recent comments).